addons/payment_paypal/controllers/main.py

```diff
@@ -46,23 +47,30 @@ def paypal_return_from_checkout(self, **pdt_data):
         retrieved and with it the transaction which will be immediately post-processed.
         """
         _logger.info("handling redirection from PayPal with data:\n%s", pprint.pformat(pdt_data))
-        if not pdt_data:  # The customer has canceled or paid then clicked on "Return to Merchant"
-            pass  # Redirect them to the status page to browse the (currently) draft transaction
-        else:
             # Check the origin of the notification
-            tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
-                'paypal', pdt_data
-            )
-            try:
-                notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
-            except Forbidden:
-                _logger.exception("could not verify the origin of the PDT; discarding it")
-            else:
+        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_notification_data(
+            'paypal', pdt_data
+        )
+        try:
+            notification_data = self._verify_pdt_notification_origin(pdt_data, tx_sudo)
+        except Forbidden:
+            _logger.exception("could not verify the origin of the PDT; discarding it")
+        else:
                 # Handle the notification data
-                tx_sudo._handle_notification_data('paypal', notification_data)
+            tx_sudo._handle_notification_data('paypal', notification_data)
 
         return request.redirect('/payment/status')
 
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET', 'POST'], csrf=False,
+        save_session=False
+    )
+    def paypal_return_from_checkout_before_paying(self, transaction_reference=None, **pdt_data):
+        # The customer has canceled payment
+        _logger.info("handling redirection for cancallation PayPal with data:\n%s", pprint.pformat(pdt_data))
+        request.env['payment.transaction'].sudo()._cancel_tx(transaction_reference)
+        return request.redirect('/payment/status')
+
```

Comment:
I thought it would be cleared since I want to pass  f'?transaction_reference={self.reference}' in url.
The third comment I don't really understand, so please elaborate, but from what I understand I have a question how would we cancel transaction if we wouldn't know which one?

Reply:
Sorry for the confusion. 

What I meant is that creating a method `_cancel_tx` in another file just for two lines seems overkill. I'm pretty sure that you can use `tx_sudo._handle_notification_data('paypal', notification_data)`, which will call the override of `_process_notification_data` in `payment_payment/model/payment_transaction.py`. By doing that, we stay in the standard flow and ensure that the behavior will stay the same for all flows. 

Finally, I think that we could reuse the return_url but I'm fine if you want to create a new route as long as you call `_handle_notification_data` in the end.

I hope that it's clearer now, but we can discuss it orally if you still have questions :)